components/object_store/src/obkv/meta.rs

```diff
@@ -0,0 +1,373 @@
+    #[serde(rename = "location")]
+    pub location: String,
+    /// The last modified time in ms
+    #[serde(rename = "last_modified")]
+    pub last_modified: i64,
+    /// The size in bytes of the object
+    #[serde(rename = "size")]
+    pub size: usize,
+    /// The unique identifier for the object; For Obkv, it is composed with
+    /// table_name @ path @ upload_id
+    #[serde(rename = "unique_id")]
+    pub unique_id: Option<String>,
+    /// The size in bytes of one part. Note: maybe the size of last part less
+    /// than part_size.
+    #[serde(rename = "part_size")]
+    pub part_size: usize,
+    /// The paths of multi upload parts.
+    #[serde(rename = "parts")]
+    pub parts: Vec<String>,
+    /// The version of object, we use the upload_id as version.
+    /// TODO: Since `upload_id` is used by multiple objects, it may become very
+    /// large. Should we assign a version number to each object to avoid
+    /// this issue?
+    #[serde(rename = "version")]
+    pub version: String,
+}
+
+impl ObkvObjectMeta {
+    #[inline]
+    pub fn decode(data: &[u8]) -> Result<Self> {
+        ensure!(
+            data[0] == HEADER,
+            InvalidHeader {
+                header: data[0],
+                expect: HEADER,
+            }
+        );
+        let json = str::from_utf8(&data[1..]).context(InvalidUtf8)?;
+        serde_json::from_str(json).context(InvalidJson { json })
+    }
+
+    #[inline]
+    pub fn encode(&self) -> Result<Vec<u8>> {
+        let size = self.estimate_size_of_json();
+        let mut encode_bytes = Vec::with_capacity(size + 1);
+        encode_bytes.push(HEADER);
+        serde_json::to_writer(&mut encode_bytes, self).context(EncodeJson)?;
+        Ok(encode_bytes)
+    }
+
+    /// Estimate the json string size of ObkvObjectMeta
+    #[inline]
+    pub fn estimate_size_of_json(&self) -> usize {
+        // {}
+        let mut size = 2;
+        // size of key name, `,`, `""` and `:`
+        size += (8 + 13 + 4 + 9 + 9 + 5 + 7) + 4 * 7;
+        size += self.location.len() + 2;
+        // last_modified
+        size += 8;
+        // size
+        size += 8;
+        // unique_id
+        if let Some(id) = &self.unique_id {
+            size += id.len() + 2;
+        } else {
+            size += 4;
+        }
+        // part_size
+        size += 8;
+        // parts
+        for part in &self.parts {
+            // part.len, `""`, `:`, and `,`
+            size += part.len() + 4;
+        }
+        //{}
+        size += 2;
+        // version
+        size += self.version.len();
+        size
+    }
+
+    /// Compute the convered parts based on given range parameter
+    pub fn compute_covered_parts(&self, range: Range<usize>) -> Result<ConveredParts> {
+        ensure!(
+            range.start <= range.end && range.end <= self.size,
+            InvalidRange {
+                start: range.start,
+                end: range.end,
+            }
+        );
```

Comment:
There is no need to check the start and end of a Range, and shall we name this error as `OutOfRange`?
```suggestion
        ensure!(
            range.end <= self.size,
            OutOfRange {
                end: range.end,
                object_size: self.size,
            }
        );
```